File: decrypt_metamask_vault.rb
#!/usr/bin/env ruby
########################################################################
# Decrypt Metamask Vault with supplied password. 
# It's broken: ruby does not seem to like 16 byte iv for AES-256 GCM
# muquit@muquit.com Nov-20-2021 
########################################################################

require 'io/console'
require 'json'
require 'base64'
require 'openssl'
require 'encryptor'

require 'pp'

class MetamaskVaultDecryptor
  def initialize
    $stdout.sync = true
    $stderr.sync = true
    @me = File.basename($0)
  end

  def log(msg)
    t = Time.new()
    puts "#{t}: #{msg}"
  end

  def log_fatal(msg)
    t = Time.new()
    puts "FATAL ERROR: #{msg}"
    exit 1
  end

  def check_args
    if ARGV.length != 1 
      puts "Usage: #{@me} <vault.json>"
      exit 1
    end
    file = ARGV[0]
    if !File.exist?(file)
      log_fatal("File #{file} does not exist")
    end
  end

  #==================================================================== 
  # It uses PBKDF2 with AES-GCM 256, SHA-256 with 10000 iteration
  # iv length is 16 bytes
  #==================================================================== 
  def key_from_password(password, salt_base64)
    log "'#{password}'"
    salt = Base64.decode64(salt_base64)
    digest = OpenSSL::Digest::SHA256.new()
    digest_len = digest.digest_length
    log "Digest length: #{digest_len}"
    key = OpenSSL::PKCS5.pbkdf2_hmac(password, salt, 10000, digest_len, digest)
    log "Key len: #{key.length}"
    keyb64 = Base64.encode64(key)
    return key
  end

  def decrypt_with_key(key, json_obj)
    cipher_text_base64 = json_obj['data']
    log "Data: #{cipher_text_base64}"
    cipher_text = Base64.decode64(cipher_text_base64)
    iv_base64 = json_obj['iv']
    log "iv: #{iv_base64}"
    iv = Base64.decode64(iv_base64)
    decipher = OpenSSL::Cipher.new('aes-256-gcm')
    decipher.decrypt
    decipher.key = key
    log "iv len: #{iv.length}"
    decipher.iv_len = iv.length
    decipher.iv = iv
    plain = decipher.update(cipher_text) + decipher.final
    pp plain
  end

  def decrypt(password, vault_json_file)
    json_str = File.read(vault_json_file)
    json = JSON.parse(json_str)
    salt_base64 = json["salt"]
    log ">> Salt: #{salt_base64}"
    key = key_from_password(password, salt_base64)
    decrypt_with_key(key, json)
  end

  def doit
    check_args()
    password = IO::console.getpass "Enter Metamask Password: "
    decrypt(password, ARGV[0])
  end

end

if __FILE__ == $0
  MetamaskVaultDecryptor.new.doit()
end

File: decrypt_metamask_vault.py
#!/usr/bin/env python3

########################################################################
#  Decrypt Metamask's encrypted vault
#  muquit@muquit.com Jan-26-2022 
########################################################################

import argparse
import logging
import sys
import json
import base64
import os, binascii
from hashlib import pbkdf2_hmac
from os.path import exists
from Crypto.Cipher import AES
from getpass import getpass

if __name__ == "__main__":
    log_format = "%(levelname)s %(asctime)s - %(message)s"
    logger = logging.getLogger(__name__)
    logging.basicConfig(
        level=logging.INFO,
        format=log_format,
        stream=sys.stdout)

    parser = argparse.ArgumentParser(description="Decrypt Metamask Vault")
    parser.add_argument('-f',
            help="Path of Metamask vault JSON file",
            required=True)
    args = parser.parse_args()
    json_file = args.f

    if exists(json_file) != True:
        print(f"Error: file {json_file} does not exist")
        exit(1)

    f = open(json_file)
    j = json.load(f)
    f.close()

    # prompt for Metamask password
    password = getpass("Enter Metamask Password: ")

    # make key from password, Metamask uses iter=10,000, key is 
    # 32 bytes long
    saltBytes = base64.b64decode(j['salt'])
    password = password.encode("utf8")
    key = pbkdf2_hmac("sha256", password, saltBytes, 10000,32)

    # decrypt data with the generated key. AES-256 GCM mode
    ivBytes = base64.b64decode(j['iv'])
    ivLen = len(ivBytes)
    dataBytes = base64.b64decode(j['data'])
    aesgcm = AES.new(key, AES.MODE_GCM, ivBytes)
    plainText = aesgcm.decrypt(dataBytes)
    print(f"{plainText}")